File: pfe-java/app/controllers/OAuth.java
package controllers;

import com.fasterxml.jackson.databind.JsonNode;
import play.libs.F;
import play.libs.Scala;
import play.libs.ws.WS;
import play.libs.ws.WSClient;
import play.mvc.Call;
import play.mvc.Controller;
import play.mvc.Http;
import play.mvc.Result;

import static controllers.URL.param;

public class OAuth extends Controller {

    public static final String TOKEN_KEY = "oauth-token";

    static final String AUTHORIZATION_ENDPOINT = "https://accounts.google.com/o/oauth2/auth";
    static final String TOKEN_ENDPOINT = "https://accounts.google.com/o/oauth2/token";
    static final String CLIENT_ID = "1079303020045-4kie53crgo06su51pi3dnbm90thc2q33.apps.googleusercontent.com";
    static final String CLIENT_SECRET = "9-PoA1ZwynHJlE4Y3VY8fONX";
    static final String SCOPE = "plus.moments.insert https://www.googleapis.com/auth/plus.login";

    static final WSClient ws = WS.client();

    public static String authorizeUrl(Call returnTo) {
        return URL.build(AUTHORIZATION_ENDPOINT, Scala.varargs(
                param("response_type", "code"),
                param("client_id", CLIENT_ID),
                param("redirect_uri", routes.OAuth.callback().absoluteURL(request())),
                param("scope", SCOPE),
                param("state", returnTo.url())
        ));
    }

    public static F.Promise<Result> callback() {
        String code = request().getQueryString("code");
        if (code != null) {
            String state = request().getQueryString("state");
            String returnTo = state != null ? state : routes.Items.list().url();
            return ws.url(TOKEN_ENDPOINT)
                    .setContentType(Http.MimeTypes.FORM)
                    .post(URL.encode(Scala.varargs(
                            param("code", code),
                            param("client_id", CLIENT_ID),
                            param("client_secret", CLIENT_SECRET),
                            param("redirect_uri", routes.OAuth.callback().absoluteURL(request())),
                            param("grant_type", "authorization_code")
                    ))).map(response -> {
                        JsonNode accessTokenJson = response.asJson().get("access_token");
                        if (accessTokenJson == null || !accessTokenJson.isTextual()) {
                            return internalServerError();
                        } else {
                            String accessToken = accessTokenJson.asText();
                            session().put(TOKEN_KEY, accessToken);
                            return redirect(returnTo);
                        }
                    });
        } else {
            return F.Promise.pure(internalServerError());
        }
    }

}

File: pfe-scala/app/controllers/OAuth.scala
package controllers

import play.api.libs.ws.WS
import play.api.mvc.{RequestHeader, Call, Action, Controller}
import play.api.libs.concurrent.Execution.Implicits.defaultContext

import scala.concurrent.Future

object OAuth extends Controller {

  val tokenKey = "oauth-token"
  
  val authorizationEndpoint = "https://accounts.google.com/o/oauth2/auth"
  val tokenEndpoint = "https://accounts.google.com/o/oauth2/token"
  val clientId = "1079303020045-4kie53crgo06su51pi3dnbm90thc2q33.apps.googleusercontent.com"
  val clientSecret = "9-PoA1ZwynHJlE4Y3VY8fONX"

  val ws = WS.client(play.api.Play.current)

  def authorizeUrl(returnTo: Call)(implicit request: RequestHeader): String =
    makeUrl(authorizationEndpoint,
        "response_type" -> "code",
        "client_id" -> clientId,
        "redirect_uri" -> routes.OAuth.callback().absoluteURL(),
        "scope" -> "https://www.googleapis.com/auth/plus.login",
        "state" -> returnTo.url
      )

  def makeUrl(endpoint: String, qs: (String, String)*): String = {
    import java.net.URLEncoder.{encode => enc}
    val params = for ((n, v) <- qs) yield s"""${enc(n, "utf-8")}=${enc(v, "utf-8")}"""
    endpoint + params.toSeq.mkString("?", "&", "")
  }

  val callback = Action.async { implicit request =>
    request.getQueryString("code") match {
      case Some(code) =>
        val returnTo = request.getQueryString("state") getOrElse routes.Items.list().url
        for {
          response <- ws.url(tokenEndpoint).post(Map(
            "code" -> Seq(code),
            "client_id" -> Seq(clientId),
            "client_secret" -> Seq(clientSecret),
            "redirect_uri" -> Seq(routes.OAuth.callback().absoluteURL()),
            "grant_type" -> Seq("authorization_code")
          ))
        } yield {
          (response.json \ "access_token").validate[String].fold(
            _ => InternalServerError,
            token => Redirect(returnTo).addingToSession(tokenKey -> token)
          )
        }
      case None =>
        Future.successful(InternalServerError)
    }
  }

}
